test/tools/ossfuzz/protoToAbiV2.cpp

```diff
@@ -0,0 +1,714 @@
+	bool _isValueType,
+	std::string const& _typeString,
+	std::string const& _varName,
+	delimiter _delimiter
+)
+{
+	std::string qualifiedTypeString = (
+		_isValueType ?
+		_typeString :
+		_typeString + " memory"
+		);
+	m_typedParamsPublic << Whiskers(R"(<delimiter><type> <varName>)")
+		("delimiter", delimiterToString(_delimiter))
+		("type", qualifiedTypeString)
+		("varName", _varName)
+		.render();
+}
+
+std::string ProtoConverter::typedParametersAsString(calleeType _calleeType)
+{
+	switch (_calleeType)
+	{
+	case calleeType::PUBLIC:
+		return m_typedParamsPublic.str();
+	case calleeType::EXTERNAL:
+		return m_typedParamsExternal.str();
+	}
+}
+
+// Function that is called by the factory contract
+void ProtoConverter::visit(TestFunction const& _x)
+{
+	m_output << Whiskers(R"(
+	function f() public returns (uint) {
+	)")
+	.render();
+
+	// Define state variables in function scope
+	m_output << m_statebuffer.str();
+
+	// TODO: Support more than one but less than N local variables
+	visit(_x.local_vars());
+
+	m_output << Whiskers(R"(
+		uint returnVal = this.g_public(<parameter_names>);
+		if (returnVal != 0)
+			return returnVal;
+		return (uint(1000) + this.g_external(<parameter_names>));
+	}
+	)")
+	("parameter_names", YulUtilFunctions::suffixedVariableNameList("x_", 0, m_varCounter))
+	.render();
+}
+
+void ProtoConverter::writeHelperFunctions()
+{
+	m_output << Whiskers(R"(
+	function stringCompare(string memory a, string memory b) internal pure returns (bool) {
+		if(bytes(a).length != bytes(b).length)
+			return false;
+		else
+			return keccak256(bytes(a)) == keccak256(bytes(b));
+	}
+	)").render();
+
+	m_output << Whiskers(R"(
+	function bytesCompare(bytes memory a, bytes memory b) internal pure returns (bool) {
+		if(a.length != b.length)
+			return false;
+		for (uint i = 0; i < a.length; i++)
+			if (a[i] != b[i])
+				return false;
+		return true;
+	}
+	)").render();
+
+	// These are callee functions that encode from storage, decode to
+	// memory/calldata and check if decoded value matches storage value
+	// return true on successful match, false otherwise
+	m_output << Whiskers(R"(
+	function g_public(<parameters_memory>) public view returns (uint) {
+		<equality_checks>
+		return 0;
+	}
+
+	function g_external(<parameters_calldata>) external view returns (uint) {
+		<equality_checks>
+		return 0;
+	}
+	)")
+	("parameters_memory", typedParametersAsString(calleeType::PUBLIC))
+	("equality_checks", equalityChecksAsString())
+	("parameters_calldata", typedParametersAsString(calleeType::EXTERNAL))
+	.render();
+}
+
+void ProtoConverter::visit(Contract const& _x)
+{
+	m_output << Whiskers(R"(pragma solidity >=0.0;
+pragma experimental ABIEncoderV2;
+
+contract Factory {
+	function test() external returns (uint) {
+		C c = new C();
+		return c.f();
+	}
+}
+
+contract C {
+)").render();
+	// TODO: Support more than one but less than N state variables
+	visit(_x.state_vars());
+	m_isStateVar = false;
```

Comment:
Why do you need to set it to false here?

Reply:
Mostly to add the `memory` qualifier for non-value types. Calls to this will return the said qualifier:

https://github.com/ethereum/solidity/pull/6880/files#diff-f3583faecac0f4d36455fb7eb07854a3R147

Reply:
Can you maybe find a better name, since we are not talking about state variables here, but more about the stage in the code generation / visiting process, is that right?

Reply:
Hmm... I was wrong. It is not just for the qualifier but also to ensure different way of handling storage types. For example, not resizing storage arrays via `new T()` but `x.length` instead; buffering assignments to storage variables etc. So it might be better to keep the name, don't you think?